src/slicing.rs

```diff
@@ -97,6 +99,14 @@ pub enum Position {
     AfterFragment
 }
 
+impl From<u32> for Position {
+    fn from(f: u32) -> Self {
+        unsafe {
+            mem::transmute(f)
```

Comment:
I think this `transmute` is unsound. This method itself is not `unsafe`, and there’s nothing stopping users from calling it with a value that doesn’t have a corresponding enum variant. This should return an option, and avoid transmute. That code would be tedious and error-prone to write by hand, but it can be generated with a macro:

```rust
macro_rules! c_like_enum {
    ( $type_name: ident: $( $variant: ident, )+ ) => {
        c_like_enum! {
            $type_name
            [ $($variant)* ]
            [ ]
            next = 0
        }
    };

    (
        $type_name: ident
        [ $next_variant: ident $( $variant: ident )* ]
        [ $($assigned_variant: ident = $value: expr),* ]
        next = $next_value: expr
    ) => {
        c_like_enum! {
            $type_name
            [ $($variant)* ]
            [ $($assigned_variant = $value,)* $next_variant = $next_value ]
            next = $next_value + 1
        }
    };

    (
        $type_name: ident
        [ ]
        [ $($variant: ident = $value: expr),+ ]
        next = $next_value: expr
    ) => {
        #[derive(Copy, Clone, Debug)]
        #[repr(u8)]
        pub enum $type_name {
            $($variant = $value,)+
        }

        impl $type_name {
            #[allow(non_upper_case_globals)]
            pub fn from_u8(value: u8) -> Option<Self> {
                $(
                    const $variant: u8 = $value;
                )+
                match value {
                    $(
                        $variant => Some($type_name::$variant),
                    )+
                    _ => None
                }
            }
        }
    };
}

c_like_enum! { Position:
    BeforeScheme,
    AfterScheme,
    BeforeUsername,
    AfterUsername,
    BeforePassword,
    AfterPassword,
    BeforeHost,
    AfterHost,
    BeforePort,
    AfterPort,
    BeforePath,
    AfterPath,
    BeforeQuery,
    AfterQuery,
    BeforeFragment,
    AfterFragment,
}